File: Payload_Type/poseidon/mythic/agent_functions/keys.py
from mythic_payloadtype_container.MythicCommandBase import *
import json


class KeysArguments(TaskArguments):
    def __init__(self, command_line, **kwargs):
        super().__init__(command_line, **kwargs)
        self.args = [
            CommandParameter(
                name="command",
                type=ParameterType.ChooseOne,
                description="Choose a way to interact with keys.",
                choices=[
                    "dumpsession",
                    "dumpuser",
                    "dumpprocess",
                    "dumpthreads",
                ],
                parameter_group_info=[
                    ParameterGroupInfo(
                        required=True
                    )
                ]
            ),
            CommandParameter(
                name="keyword",
                type=ParameterType.String,
                description="Name of the key to search for",
                parameter_group_info=[
                    ParameterGroupInfo(
                        required=True,
                        group_name="search"
                    )
                ]
            ),
            CommandParameter(
                name="typename",
                type=ParameterType.ChooseOne,
                description="Choose the type of key",
                choices=["keyring", "user", "login", "logon", "session"],
                parameter_group_info=[
                    ParameterGroupInfo(
                        required=True,
                        group_name="search"
                    )
                ]
            ),
        ]

    async def parse_arguments(self):
        self.load_args_from_json_string(self.command_line)

    async def parse_dictionary(self, dictionary):
        self.load_args_from_dictionary(dictionary)
        if self.get_parameter_group_name() == "search":
            self.remove_arg("command")
            self.add_arg("command", value="search", parameter_group_info=[ParameterGroupInfo(group_name="search")])


class KeysCommand(CommandBase):
    cmd = "keys"
    needs_admin = False
    help_cmd = "keys"
    description = "Interact with the linux keyring."
    version = 1
    author = "@xorrior"
    argument_class = KeysArguments
    attackmapping = []
    attributes = CommandAttributes(
        supported_os=[SupportedOS.Linux]
    )

    async def create_tasking(self, task: MythicTask) -> MythicTask:
        return task

    async def process_response(self, response: AgentResponse):
        pass

File: Payload_Type/poseidon/poseidon/agentfunctions/keys.go
package agentfunctions

import (
	agentstructs "github.com/MythicMeta/MythicContainer/agent_structs"
)

func init() {
	agentstructs.AllPayloadData.Get("poseidon").AddCommand(agentstructs.Command{
		Name:                "keys",
		HelpString:          "keys",
		Description:         "Interact with the linux keyring",
		Version:             1,
		MitreAttackMappings: []string{},
		Author:              "@xorrior",
		CommandParameters: []agentstructs.CommandParameter{
			{
				Name:          "command",
				ParameterType: agentstructs.COMMAND_PARAMETER_TYPE_CHOOSE_ONE,
				Description:   "Choose a way to interact with the keyring",
				ParameterGroupInformation: []agentstructs.ParameterGroupInfo{
					{
						ParameterIsRequired: true,
						UIModalPosition:     1,
					},
				},
				Choices: []string{"dumpsession", "dumpuser", "dumpprocess", "dumpthreads"},
			},
			{
				Name:          "keyword",
				Description:   "Name of the key to search for",
				ParameterType: agentstructs.COMMAND_PARAMETER_TYPE_STRING,
				ParameterGroupInformation: []agentstructs.ParameterGroupInfo{
					{
						ParameterIsRequired: true,
						UIModalPosition:     2,
						GroupName:           "search",
					},
				},
			},
			{
				Name:          "typename",
				ParameterType: agentstructs.COMMAND_PARAMETER_TYPE_CHOOSE_ONE,
				Description:   "Choose the type of the key",
				ParameterGroupInformation: []agentstructs.ParameterGroupInfo{
					{
						ParameterIsRequired: true,
						UIModalPosition:     3,
						GroupName:           "search",
					},
				},
				Choices: []string{"keyring", "user", "login", "login", "session"},
			},
		},
		CommandAttributes: agentstructs.CommandAttribute{
			SupportedOS: []string{agentstructs.SUPPORTED_OS_LINUX},
		},
		TaskFunctionParseArgString: func(args *agentstructs.PTTaskMessageArgsData, input string) error {
			if err := args.LoadArgsFromJSONString(input); err != nil {
				return err
			} else if groupName, err := args.GetParameterGroupName(); err != nil {
				return err
			} else if groupName == "search" {
				args.RemoveArg("command")
				args.AddArg(agentstructs.CommandParameter{
					Name:         "command",
					DefaultValue: "search",
					ParameterGroupInformation: []agentstructs.ParameterGroupInfo{
						{
							GroupName: "search",
						},
					},
				})
				return nil
			} else {
				return nil
			}
		},
		TaskFunctionParseArgDictionary: func(args *agentstructs.PTTaskMessageArgsData, input map[string]interface{}) error {
			if err := args.LoadArgsFromDictionary(input); err != nil {
				return err
			} else if groupName, err := args.GetParameterGroupName(); err != nil {
				return err
			} else if groupName == "search" {
				args.RemoveArg("command")
				args.AddArg(agentstructs.CommandParameter{
					Name:         "command",
					DefaultValue: "search",
					ParameterGroupInformation: []agentstructs.ParameterGroupInfo{
						{
							GroupName: "search",
						},
					},
				})
				return nil
			} else {
				return nil
			}
		},
		TaskFunctionCreateTasking: func(task *agentstructs.PTTaskMessageAllData) agentstructs.PTTaskCreateTaskingMessageResponse {
			response := agentstructs.PTTaskCreateTaskingMessageResponse{
				Success: true,
				TaskID:  task.Task.ID,
			}
			return response
		},
	})
}
